Readme.txt
Original Readme.txt can (and should) be found as: Readme-original.txt
Please read it!

From the original documentation:

"USBasp is a USB in-circuit programmer for Atmel AVR controllers. It simply
 consists of an ATMega88 or an ATMega8 and a couple of passive components.
 The programmer uses a firmware-only USB driver, no special USB controller
 is needed."

This repository contains a patched version of the original USBasp firmware
compared to the "official" firmware. Please note, that you may have problems
with this one, use this at your OWN risk! If you are a regular user not so
interested in extra features, please stick with the official firmware!

The official firmware can be found here: http://www.fischl.de/usbasp/

History, important changes compared to the original firmware:

Step 1:

	http://www.fischl.de/usbasp/
	usbasp.2011-05-28.tar.gz
	Original/official firmware
	Readme-original.txt contains the original Readme of the project
	Changelog-original.txt contains the original Changelog of the project

Step 2:

	https://github.com/stefanbeller/USBasp
	Original firmware was imported to a Github respository by Stefan
	Beller. It was then modified/extended by him, key features:

	* Missing "const" keywords at "PROGMEM" directives to allow to compile
	  the source with newer avr-gcc versions (older ones seems to ignore
	  the problem so it worked)

	* Target pushed data (as target being the SPI master) via SPI is stored
	  in the communication buffer, which can be read from your PC. A simple
	  Python based tool is provided by him for this purpose. It's only
	  uni-directional

	Greetings, great work of the author, thanks!

Step 3:

	http://www.avrfreaks.net/projects/usbasp-tty-usbasp-programmer-modified-serial-support-and-terminal-program
	"Emklaus" (aka EMK) patched the original (step 1) firmware to
	allow to use the RX/TX (TTL serial) lines of the programmer, so the
	target device can communicate with the PC through the USBasp which
	"relays" (bi-directional) the information towards the PC. A Windows
	based client software is also provided in the archive. Currently I
	did not import the client software itself, as it's too platform
	dependent (Windows only, I do not use Windows).
	EMK's work often referred as "USBasp-TTY".

	Greetings, great work of the author, thanks!

Step 4:

	https://github.com/lgblgblgb/USBasp
	I (LGB, lgblgblgb on github) have forked Stefan Beller's repository on
	Github (step 2), and merged EMK's (step 3) changes into the repository
	with some manual work (it wouldn't apply automatically by "patch").
	Also, some cosmetical changes, and unify line endings hell (mixed
	DOS / UNIX style). Of course much more work is planned for the future.
	Changelog.txt contains the changelog of my changes (including my work
	and using other's work as well, as it noted above)

Step 5:

	Plans, ongoing new work from me, testing.

	Including:

	utils/usbasp.py will be a python based utility with functions including
	pipeout.py's but also USBasp-tty client works later.



Changelog.txt
2015-Sep-14
- many cosmetic/directory structure/documentation changes
- dropping windows driver, I don't think it belongs here too much ...
- V-USB update from its repository

2015-Sep-13
- EMK's USBasp-tty changes are merged
  http://www.avrfreaks.net/projects/usbasp-tty-usbasp-programmer-modified-serial-support-and-terminal-program



firmware/main.c
/*
 * USBasp - USB in-circuit programmer for Atmel AVR controllers
 *
 * Thomas Fischl <tfischl@gmx.de>
 *
 * License........: GNU GPL v2 (see Readme.txt)
 * Target.........: ATMega8 at 12 MHz
 * Creation Date..: 2005-02-20
 * Last change....: 2009-02-28
 *
 * PC2 SCK speed option.
 * GND  -> slow (8khz SCK),
 * open -> software set speed (default is 375kHz SCK)
 */

#include <avr/io.h>
#include <avr/interrupt.h>
#include <avr/pgmspace.h>
#include <avr/wdt.h>

#include "usbasp.h"
#include "usbdrv.h"
#include "isp.h"
#include "clock.h"
#include "tpi.h"
#include "tpi_defs.h"
#include "uart.h"

static uchar replyBuffer[8];

// choose size to 2^8, so comStart, comEnd can wrap around as ringbuffer
// indexes.
static uchar comBuffer[256];
static uchar comStart = 0;
static uchar comStop = 0;

unsigned int blink_counter;

static uchar prog_state = PROG_STATE_IDLE;
static uchar prog_sck = USBASP_ISP_SCK_AUTO;

static uchar prog_address_newmode = 0;
static unsigned long prog_address;
static unsigned int prog_nbytes = 0;
static unsigned int prog_pagesize;
static uchar prog_blockflags;
static uchar prog_pagecounter;
static uchar debug_byte = 'X';
static unsigned long baud = BAUD_RATE;

ISR(SPI_STC_vect, ISR_NOBLOCK)
{
    comBuffer[comStop] = SPDR;
    comStop ++;
}

uchar usbFunctionSetup(uchar data[8]) {

	uchar len = 0, i;

	if (data[1] == USBASP_FUNC_CONNECT) {

		/* set SCK speed */
		if ((PINC & (1 << PC2)) == 0) {
			ispSetSCKOption(USBASP_ISP_SCK_8);
		} else {
			ispSetSCKOption(prog_sck);
		}

		/* set compatibility mode of address delivering */
		prog_address_newmode = 0;

		ledRedOn();
		ispConnect();

	} else if (data[1] == USBASP_FUNC_DISCONNECT) {
		ispDisconnect();
		ledRedOff();

	} else if (data[1] == USBASP_FUNC_TRANSMIT) {
		replyBuffer[0] = ispTransmit(data[2]);
		replyBuffer[1] = ispTransmit(data[3]);
		replyBuffer[2] = ispTransmit(data[4]);
		replyBuffer[3] = ispTransmit(data[5]);
		len = 4;

	} else if (data[1] == USBASP_FUNC_READFLASH) {

		if (!prog_address_newmode)
			prog_address = (data[3] << 8) | data[2];

		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_READFLASH;
		len = 0xff; /* multiple in */

	} else if (data[1] == USBASP_FUNC_READEEPROM) {

		if (!prog_address_newmode)
			prog_address = (data[3] << 8) | data[2];

		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_READEEPROM;
		len = 0xff; /* multiple in */

	} else if (data[1] == USBASP_FUNC_ENABLEPROG) {
		replyBuffer[0] = ispEnterProgrammingMode();
		len = 1;

	} else if (data[1] == USBASP_FUNC_WRITEFLASH) {

		if (!prog_address_newmode)
			prog_address = (data[3] << 8) | data[2];

		prog_pagesize = data[4];
		prog_blockflags = data[5] & 0x0F;
		prog_pagesize += (((unsigned int) data[5] & 0xF0) << 4);
		if (prog_blockflags & PROG_BLOCKFLAG_FIRST) {
			prog_pagecounter = prog_pagesize;
		}
		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_WRITEFLASH;
		len = 0xff; /* multiple out */

	} else if (data[1] == USBASP_FUNC_WRITEEEPROM) {

		if (!prog_address_newmode)
			prog_address = (data[3] << 8) | data[2];

		prog_pagesize = 0;
		prog_blockflags = 0;
		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_WRITEEEPROM;
		len = 0xff; /* multiple out */

	} else if (data[1] == USBASP_FUNC_SETLONGADDRESS) {

		/* set new mode of address delivering (ignore address delivered in commands) */
		prog_address_newmode = 1;
		/* set new address */
		prog_address = *((unsigned long*) &data[2]);

	} else if (data[1] == USBASP_FUNC_SETISPSCK) {

		/* set sck option */
		prog_sck = data[2];
		replyBuffer[0] = 0;
		len = 1;

	} else if (data[1] == USBASP_FUNC_TPI_CONNECT) {
		tpi_dly_cnt = data[2] | (data[3] << 8);

		/* RST high */
		ISP_OUT |= (1 << ISP_RST);
		ISP_DDR |= (1 << ISP_RST);

		clockWait(3);

		/* RST low */
		ISP_OUT &= ~(1 << ISP_RST);
		ledRedOn();

		clockWait(16);
		tpi_init();

	} else if (data[1] == USBASP_FUNC_TPI_DISCONNECT) {

		tpi_send_byte(TPI_OP_SSTCS(TPISR));
		tpi_send_byte(0);

		clockWait(10);

		/* pulse RST */
		ISP_OUT |= (1 << ISP_RST);
		clockWait(5);
		ISP_OUT &= ~(1 << ISP_RST);
		clockWait(5);

		/* set all ISP pins inputs */
		ISP_DDR &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));
		/* switch pullups off */
		ISP_OUT &= ~((1 << ISP_RST) | (1 << ISP_SCK) | (1 << ISP_MOSI));

		ledRedOff();

	} else if (data[1] == USBASP_FUNC_TPI_RAWREAD) {
		replyBuffer[0] = tpi_recv_byte();
		len = 1;

	} else if (data[1] == USBASP_FUNC_TPI_RAWWRITE) {
		tpi_send_byte(data[2]);

	} else if (data[1] == USBASP_FUNC_TPI_READBLOCK) {
		prog_address = (data[3] << 8) | data[2];
		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_TPI_READ;
		len = 0xff; /* multiple in */

	} else if (data[1] == USBASP_FUNC_TPI_WRITEBLOCK) {
		prog_address = (data[3] << 8) | data[2];
		prog_nbytes = (data[7] << 8) | data[6];
		prog_state = PROG_STATE_TPI_WRITE;
		len = 0xff; /* multiple out */

	} else if (data[1] == USBASP_FUNC_GETCAPABILITIES) {
		replyBuffer[0] = USBASP_CAP_0_TPI;
		replyBuffer[1] = 0;
		replyBuffer[2] = 0;
		replyBuffer[3] = 0;
		len = 4;

	} else if (data[1] == USBASP_FUNC_SPI_RECVSTART) {
		spiInit();
		prog_state = PROG_STATE_SERIAL;

	} else if (data[1] == USBASP_FUNC_SPI_RECVSTOP) {
		ledRedOff();
		ispDisconnect();
		prog_state = PROG_STATE_IDLE;

	} else if (data[1] == USBASP_FUNC_SPI_RECV) {
		len = 0;
		while (comStart != comStop && len < 8) {
			replyBuffer[len] = comBuffer[comStart];
			comStart ++;
			len ++;
		}
	/* ---- EMK Modifications for DEBUG serial interface ---- */
	} else if (data[1] == USBASP_FUNC_UART_GETBYTE) {
		/* Get a byte From UART outBuf (if any) send to HOST */
		if (UART_outbufsize() > 0)
			replyBuffer[0] = UART_outBuf_get();	/* get a byte from the outBuf buffer */
		else
			replyBuffer[0] = 0;	/* nothing to send, send NULL byte */
		len = 1;
	} else if (data[1] == USBASP_FUNC_UART_PUTBYTE) {
		/* Get Data From HOST to be sent to UART */
		debug_byte = data[2];		/* save byte to debug_byte */
		PORTC &= 0xFC;			/* clear Low 2 bits */
		PORTC |= ~debug_byte & 0x03;	/* show Low 2 bits in LEDs */
		UART_inBuf_put(debug_byte);	/* save in inBuf for sending to UART */
	} else if (data[1] == USBASP_FUNC_UART_GETBYTECOUNT) {
		/* Get #of bytes from UART waiting to be sent to HOST */
		replyBuffer[0] = UART_outbufsize();  /* get #of bytes in outBuf buffer */
		len = 1;
	} else if (data[1] == USBASP_FUNC_UART_SETBAUDRATE) {
		/* Set UART Baud Rate (53) */
		baud = *((unsigned long*) & data[2]);
		UART_init(baud);
		replyBuffer[0] = 0x53;	/* return Code = command in Hex */
		len = 1;
	} else if (data[1] == USBASP_FUNC_TEST_CMD1) {
		/* Just For Demonstration ... */
		PORTC |= 0x03;	/* both LEDs OFF */
		for(i = 0; i < 10; i++) {	/* make the LEDs blink for 816ms */
			clockWait(255);		/* 320us * 255 = 81.6ms */
			PORTC++;
		}
		PORTC = 0xFC;			/* restore the state of the LEDs */
		replyBuffer[0] = debug_byte;	/* just return last byte sent from Host */
		len = 1;
	}
	/* [END] EMK Modifications for DEBUG serial interface */

	usbMsgPtr = replyBuffer;

	return len;
}

uchar usbFunctionRead(uchar *data, uchar len) {

	uchar i;

	/* check if programmer is in correct read state */
	if ((prog_state != PROG_STATE_READFLASH) && (prog_state
			!= PROG_STATE_READEEPROM) && (prog_state != PROG_STATE_TPI_READ)) {
		return 0xff;
	}

	/* fill packet TPI mode */
	if(prog_state == PROG_STATE_TPI_READ)
	{
		tpi_read_block(prog_address, data, len);
		prog_address += len;
		return len;
	}

	/* fill packet ISP mode */
	for (i = 0; i < len; i++) {
		if (prog_state == PROG_STATE_READFLASH) {
			data[i] = ispReadFlash(prog_address);
		} else {
			data[i] = ispReadEEPROM(prog_address);
		}
		prog_address++;
	}

	/* last packet? */
	if (len < 8) {
		prog_state = PROG_STATE_IDLE;
	}

	return len;
}

uchar usbFunctionWrite(uchar *data, uchar len) {

	uchar retVal = 0;
	uchar i;

	/* check if programmer is in correct write state */
	if ((prog_state != PROG_STATE_WRITEFLASH) && (prog_state
			!= PROG_STATE_WRITEEEPROM) && (prog_state != PROG_STATE_TPI_WRITE)) {
		return 0xff;
	}

	if (prog_state == PROG_STATE_TPI_WRITE)
	{
		tpi_write_block(prog_address, data, len);
		prog_address += len;
		prog_nbytes -= len;
		if(prog_nbytes <= 0)
		{
			prog_state = PROG_STATE_IDLE;
			return 1;
		}
		return 0;
	}

	for (i = 0; i < len; i++) {

		if (prog_state == PROG_STATE_WRITEFLASH) {
			/* Flash */

			if (prog_pagesize == 0) {
				/* not paged */
				ispWriteFlash(prog_address, data[i], 1);
			} else {
				/* paged */
				ispWriteFlash(prog_address, data[i], 0);
				prog_pagecounter--;
				if (prog_pagecounter == 0) {
					ispFlushPage(prog_address, data[i]);
					prog_pagecounter = prog_pagesize;
				}
			}

		} else {
			/* EEPROM */
			ispWriteEEPROM(prog_address, data[i]);
		}

		prog_nbytes--;

		if (prog_nbytes == 0) {
			prog_state = PROG_STATE_IDLE;
			if ((prog_blockflags & PROG_BLOCKFLAG_LAST) && (prog_pagecounter
					!= prog_pagesize)) {

				/* last block and page flush pending, so flush it now */
				ispFlushPage(prog_address, data[i]);
			}

			retVal = 1; // Need to return 1 when no more data is to be received
		}

		prog_address++;
	}

	return retVal;
}

int main(void) {
	uchar i, j;

	/* no pullups on USB and ISP pins */
	PORTD = 0;
	PORTB = 0;
	/* all outputs except PD2 = INT0 */
	DDRD = ~(1 << 2);

	/* output SE0 for USB reset */
	DDRB = ~0;
	j = 0;
	/* USB Reset by device only required on Watchdog Reset */
	while (--j) {
		i = 0;
		/* delay >10ms for USB reset */
		while (--i)
			;
	}
	/* all USB and ISP pins inputs */
	DDRB = 0;

	/* all inputs except PC0, PC1 */
	DDRC = 0x03;
	PORTC = 0xfe;

	/* init UART */
	UART_init(baud);

	/* init timer */
	clockInit();

	/* main event loop */
	usbInit();

#ifdef TEST_TARGET_AS_SPI_MASTER
	DDRB = (1 << PB2 | 1 << PB3 | 1 << PB5);
	PORTB &= (1 << PB2);
	SPCR = (1 << SPE | 1 << MSTR);
	sei();
	for (;;) {
		clockWait(30); // wait 9,6 ms
		toggleLedRed();
		SPDR = j++;
		clockWait(1); // wait 320 us
		SPDR = j++;
		if (j >= 'Z')
			j = 'A';
	}
#endif

	sei();
	for (;;) {
		usbPoll();
		if (prog_state == PROG_STATE_SERIAL) {
			if (--blink_counter == 0) {
				toggleLedRed();
				blink_counter = (unsigned int)(1UL << 15);
			}
		} else
			UART_poll();
	}
	return 0;
}



firmware/uart.h
#ifndef UART_H
#define UART_H

// INBUF_SIZE & OUTBUFSIZE MUST be power of 2 (e.g. 2,4,8,16,32,64,128...)
#define UART_INBUF_SIZE          128
#define UART_OUTBUF_SIZE         128
#define UART_INBUF_MASK          UART_INBUF_SIZE-1
#define UART_OUTBUF_MASK         UART_OUTBUF_SIZE-1

/* ** NOTE: These macros may need to be adjusted for your device ** */
#define UART_inp_ready() (UCSRA & _BV(RXC))!=0
#define UART_out_ready() (UCSRA & _BV(UDRE))!=0

// ***********************************************************
// UART_init(int baud)
//  Set Up USART   NOT Interrupt drived
// ***********************************************************
void UART_init(unsigned long baud);

// ***********************************************************
// UART_get_chr(void)
//  Wait for a character to be recieved and return it's value
// ***********************************************************
char UART_get_chr(void);

// ***********************************************************
//  UART_put_chr(char c)
//  Wait for output buffer to be empty then send the character.
// ***********************************************************
void UART_put_chr(char c);

// ***********************************************************
//  UART_put_str(char *str)
//  Send characters to USART (str) until NULL is encountered.
// ***********************************************************
void UART_put_str(char *str);

// ***********************************************************
//  UART_poll()
//  Process input and output data Host to UART and UART to Host
// ***********************************************************
void UART_poll(void);


// ***********************************************************
//  UART_calc_bufsize(head, tail, size)
//  calculate the number of bytes in the specified bufer (inBuf or outBuf)
// ***********************************************************
unsigned char UART_calc_bufsize(unsigned char head, unsigned char tail, unsigned char size);

// ***********************************************************
//  UART_outbufsize()
//  return the number of bytes in outBuf
// ***********************************************************  
unsigned char UART_outbufsize(void);

// ***********************************************************
//  UART_inbufsize()
//  return the number of bytes in inBuf
// ***********************************************************  
unsigned char UART_inbufsize(void);

// ***********************************************************
//  UART_inBuf_put(c)
//  Add a byte to InBuf and adjust pointer
// ***********************************************************
void UART_inBuf_put(unsigned char c);

// ***********************************************************
//  UART_outBuf_put(c)
//  Add a byte to outBuf and adjust pointer 
// ***********************************************************
void UART_outBuf_put(unsigned char c);

// ***********************************************************
//  UART_inBuf_get()
//  Get a byte from inBuf and adjust pointer
// ***********************************************************
unsigned char UART_inBuf_get(void);

// ***********************************************************
//  UART_outBuf_get()
//  Get a byte from outBuf and adjust pointer
// ***********************************************************
unsigned char UART_outBuf_get(void);

#endif



utils/pipeout.py
#!/usr/bin/env python

import usb
import time
import sys

def getHandle():
    asp = None
    busses = usb.busses()
    for bus in busses:
        devices = bus.devices
        for dev in devices:
            if dev.idVendor == 0x16c0 and dev.idProduct==0x05dc:
                asp = dev #.open()
    if asp is None:
        return None
    return asp.open()

def enableSerialMode(USBaspHandle):
    USBaspHandle.controlMsg(128|32, 32, 8)

def disableSerialMode(USBaspHandle):
    USBaspHandle.controlMsg(128|32, 34, 8)

def getData(USBaspHandle):
    return USBaspHandle.controlMsg(128|32, 33, 8)

handle = getHandle()
if handle is None:
    sys.stderr.write("USBASP not found!\n")
    exit(1);

try:
    enableSerialMode(handle)
    while True:
        data = getData(handle)
        time.sleep(0.01)
        if len(data):
            for char in data:
                sys.stdout.write(chr(char))
            sys.stdout.flush()
except KeyboardInterrupt:
    disableSerialMode(handle)
except:
    disableSerialMode(handle)
    raise


firmware/usbasp.h
/*
 * usbasp.c - part of USBasp
 *
 * Autor..........: Thomas Fischl <tfischl@gmx.de>
 * Description....: Definitions and macros for usbasp
 * Licence........: GNU GPL v2 (see Readme.txt)
 * Creation Date..: 2009-02-28
 * Last change....: 2009-02-28
 */

#ifndef USBASP_H_
#define USBASP_H_

/* USB function call identifiers */
#define USBASP_FUNC_CONNECT     1
#define USBASP_FUNC_DISCONNECT  2
#define USBASP_FUNC_TRANSMIT    3
#define USBASP_FUNC_READFLASH   4
#define USBASP_FUNC_ENABLEPROG  5
#define USBASP_FUNC_WRITEFLASH  6
#define USBASP_FUNC_READEEPROM  7
#define USBASP_FUNC_WRITEEEPROM 8
#define USBASP_FUNC_SETLONGADDRESS 9
#define USBASP_FUNC_SETISPSCK 10
#define USBASP_FUNC_TPI_CONNECT      11
#define USBASP_FUNC_TPI_DISCONNECT   12
#define USBASP_FUNC_TPI_RAWREAD      13
#define USBASP_FUNC_TPI_RAWWRITE     14
#define USBASP_FUNC_TPI_READBLOCK    15
#define USBASP_FUNC_TPI_WRITEBLOCK   16

#define USBASP_FUNC_SPI_RECVSTART    32
#define USBASP_FUNC_SPI_RECV         33
#define USBASP_FUNC_SPI_RECVSTOP     34


/* EMK Modifications for DEBUG serial interface */
#define USBASP_FUNC_UART_PUTBYTE      50
#define USBASP_FUNC_UART_GETBYTE      51
#define USBASP_FUNC_UART_GETBYTECOUNT 52
#define USBASP_FUNC_UART_SETBAUDRATE  53

#define USBASP_FUNC_TEST_CMD1         61
#define USBASP_FUNC_TEST_CMD2         62
#define USBASP_FUNC_TEST_CMD3         63
/* [END]EMK Modifications for DEBUG serial interface */

#define USBASP_FUNC_GETCAPABILITIES 127

/* EMK Modifications for DEBUG serial interface */
/* UART Baud Rate */
#define BAUD_RATE 9600
/* [END]EMK Modifications for DEBUG serial interface */

/* USBASP capabilities */
#define USBASP_CAP_0_TPI    0x01

/* programming state */
#define PROG_STATE_IDLE         0
#define PROG_STATE_WRITEFLASH   1
#define PROG_STATE_READFLASH    2
#define PROG_STATE_READEEPROM   3
#define PROG_STATE_WRITEEEPROM  4
#define PROG_STATE_TPI_READ     5
#define PROG_STATE_TPI_WRITE    6
#define PROG_STATE_SERIAL       7

/* Block mode flags */
#define PROG_BLOCKFLAG_FIRST    1
#define PROG_BLOCKFLAG_LAST     2

/* ISP SCK speed identifiers */
#define USBASP_ISP_SCK_AUTO   0
#define USBASP_ISP_SCK_0_5    1   /* 500 Hz */
#define USBASP_ISP_SCK_1      2   /*   1 kHz */
#define USBASP_ISP_SCK_2      3   /*   2 kHz */
#define USBASP_ISP_SCK_4      4   /*   4 kHz */
#define USBASP_ISP_SCK_8      5   /*   8 kHz */
#define USBASP_ISP_SCK_16     6   /*  16 kHz */
#define USBASP_ISP_SCK_32     7   /*  32 kHz */
#define USBASP_ISP_SCK_93_75  8   /*  93.75 kHz */
#define USBASP_ISP_SCK_187_5  9   /* 187.5  kHz */
#define USBASP_ISP_SCK_375    10  /* 375 kHz   */
#define USBASP_ISP_SCK_750    11  /* 750 kHz   */
#define USBASP_ISP_SCK_1500   12  /* 1.5 MHz   */

/* macros for gpio functions */
#define ledRedOn()    PORTC &= ~(1 << PC1)
#define ledRedOff()   PORTC |= (1 << PC1)
#define ledGreenOn()  PORTC &= ~(1 << PC0)
#define ledGreenOff() PORTC |= (1 << PC0)

#define isLedRedOff()   (PORTC & (1 << PC1))
#define isLedRedOn()    ~(isLedRedOff())
#define isLedGreenOff() (PORTC & (1 << PC0))
#define isLedGreenOn()  ~(isLedGreenOff())

#define toggleLedRed()  PORTC ^= (1 << PC1)
#define toggleLedGreen()  PORTC ^= (1 << PC0)


#endif /* USBASP_H_ */
